websites/management/commands/markdown_cleaning/rules/broken_link_fix.py

```diff
@@ -0,0 +1,284 @@
+
+        replacement, notes = result.original_text, Notes()
+
+        if found_wc and url_path.endswith(("_index", "index.htm", "index.html")):
+            # Remove the extra _index* url component.
+            replacement, fix_note = self.create_replacement(result, url, found_wc)
+            notes = Notes(
+                issue_type="_index postfix",
+                fix=fix_note,
+                linked_content=found_wc.text_id,
+            )
+        elif found_wc:
+            # A WebsiteContent was successfully discovered. We probably don't
+            # need to do anything.
+            replacement, notes = result.original_text, Notes(issue_type="Nothing to do")
+        else:
+            # Start making guesses.
+            found_wc = self._find_best_matching_content(url, wc)
+            notes = Notes(issue_type="Unknown link")
+
+            if found_wc:
+                replacement, fix_note = self.create_replacement(result, url, found_wc)
+                notes = Notes(
+                    issue_type="Unknown link",
+                    fix=fix_note,
+                    linked_content=found_wc.text_id,
+                )
+
+        return replacement, notes
+
+
+class BrokenMarkdownLinkFixRule(BrokenLinkFixRuleMixin, PyparsingRule):
+    """
+    A rule to fix broken links in markdown.
+
+    This rule replaces broken links with shortcodes (i.e. resource_link or resource).
+
+    For information about the types of links replaced, see BrokenLinkFixRuleMixin.
+    """
+
+    alias = "broken_markdown_link_fix"
+
+    fields = [
+        "markdown",
+    ]
+
+    def create_replacement(
+        self, result: LinkParseResult, url: ParseResult, wc: WebsiteContent
+    ) -> (str, str):
+        """
+        Return a shortcode replacement text for `wc`.
+
+        Returns:
+            tuple(str, str): The shortcode text and note/comment.
+        """
+        try:
+            if result.link.is_image:
+                sc = ShortcodeTag.resource(wc.text_id)
+            else:
+                sc = ShortcodeTag.resource_link(
+                    wc.text_id, result.link.text, url.fragment
+                )
+            return sc.to_hugo(), "Replaced with shortcode"
+        except:  # noqa: E722
+            return result.original_text, ""
+
+
+class BrokenMetadataLinkFixRule(BrokenLinkFixRuleMixin, PyparsingRule):
+    """
+    Fix links.
+    """
+
+    alias = "broken_metadata_link_fix"
+
+    fields = [
+        "metadata.related_resources_text",
+        "metadata.image_metadata.caption",
+        "metadata.image_metadata.credit",
+        "metadata.optional_text",
+        "metadata.description",
+        "metadata.course_description",
+    ]
+
+    def create_replacement(
+        self, result: LinkParseResult, url: ParseResult, wc: WebsiteContent
+    ) -> (str, str):
+        """
+        Return a markdown link replacement text for `wc` (because
+        shortcodes are not supported in metadata yet.)
```

Comment:
This was true when I worked on markdown cleanup in March 2022, but it's not entirely true now: if metadata content is rendered with `renderString`, then shortcodes will be rendered.

https://github.com/gohugoio/hugo/issues/6703

I believe we generally use `renderString` to render metadata content, e.g., course description.

Reply:
I see. That's awesome. Thanks for the info.

I will change the doc. Would you also like me to change the replacement text? I looked at Studio and don't think authors can edit shortcodes in metadata fields yet.

Reply:
>  I looked at Studio and don't think authors can edit shortcodes in metadata fields yet.

Good point.

I don't think you need to change the replacement text. Converting metadata links to shortcodes is something we should consider, but probably best handled separately.